draft-cpbs-pseudorandom-ctls.md

```diff
@@ -66,7 +66,9 @@ The goal of this extension is to enable two endpoints to agree on a TLS-based pr
 * Privacy: A third party without access to the template cannot tell whether two connections are using the same pseudorandom cTLS template, or two different pseudorandom cTLS templates.
 * Ossification risk: Every byte sent on the underlying transport is pseudorandom to an observer who does not know the cTLS template.
 * Efficiency: Zero size overhead and minimal CPU cost.  Support for servers with many cTLS templates, when appropriately constructed.
-* Protocol confusion attack resistance: The sender can ensure that the wire image does not deviate substantially from pseudorandom, even if the plaintext is controlled by an attacker who knows all the secrets.
+* Protocol confusion attack resistance: This attack assumes a malicious server or client that can coerce its peer into sending particular plaintext, in order to produce ciphertext that could be misinterpreted as a different protocol by a third party.  This extension must enable each peer to ensure that its own output is pseudorandom or nearly-pseudorandom in such a scenario.
```

Comment:
This is much better, but: nit: "pseudorandom" is still a bit confusing. How about "This extension must make enable each peer to ensure that its output is unpredictable to its peer."

Reply:
Adjusted to "This extension must enable each peer to ensure that its own output is unlikely to resemble any other protocol.".

> > > Straw-man: What about STPRP-enciphering the AEAD-encrypted records? This is less efficient, obviously, but specifying this would be less complex. It also allows us to focus the analysis one primitive rather than two. And besides, we're already conceding a loss in efficiency in order to provide resistance to protocol-confusion attacks.
> 
> > Unfortunately, this is not sufficient.  STPRPs have ~zero generic guarantees against an attacker who holds the private key.  As an example, HCTR2(AES-GCM(...)) is presumably not sufficient, because HCTR2 also relies on a GMAC-like construction, so the above attack can likely be applied twice.
> 
> I agree that's not sufficient in general. We think we're going to have to model the STPRP as an ideal cipher; some schemes will be suitable for this, others won't.

I'm not aware of any well-known STPRPs that could be modeled this way, and I'm hesitant to start trying to classify them or invent one.  Inventing a new AEAD (as here) actually seems easier (and more computationally efficient).

> That's what (in)differenitability aims to capture. For some STPRPs (resp. AEADs) we'll be able to prove that the scheme is indifferentiable from an ideal cipher (ideal AE) when we model some underlying primitive (HMAC, say) as an RO.

Specifically, I haven't encountered any STPRPs that are built around hash functions, although I'm sure it's been done.

> > Adding entropy to the associated data doesn't work in general.  For example, in AES-GCM, the associated data only affects the MAC value.  I see four options:
> > 
> > * Randomize the nonce.  This creates the possibility of nonce reuse due to collisions, so it might require lowering the cipher use limits for some ciphers.
> > * Randomize the key.  This requires rerunning the key schedule for each message.  AES-GCM-SIV does this so I guess it's not prohibitive, but it's not free either.  Probably safe... EDIT: Except for related-key attacks, which seem like a significant concern.
> > * Use a randomized cipher suite.  These already exist for TLS 1.2 (the CBC cipher suites, which use a random IV), but not for TLS 1.3.  I'm not aware of any randomized AEADs that also take a nonce, so we would need to define a new mode.
> > * Add entropy _and_ layering.  Re-encrypting the ciphertext with any freshly keyed cipher (even a stream cipher) would work.  Wrapping (ciphertext + noise) in an STPRP is also probably sufficient, although it requires a stronger model.
> 
> Options 1 and 3 don't seem feasible without invasive changes to TLS 1.3, changes which would require revisiting existing security analysis.

I assume you mean 1 and 2.  I agree, although maybe re-running the HKDF with extra entropy to generate a new key for each message would be safe.

> Option 3 doesn't seem to solve the problem, unless the attacker doesn't know which key the peer will derive. In order for this to work the peer would have to add fresh entropy to the key schedule, which the attacker would also be able to compute.

Yes, TLS 1.2 cipher suites like TLS_RSA_WITH_AES_128_CBC_SHA (the default!) include a fresh random "IV" in [every record header](https://datatracker.ietf.org/doc/html/rfc5246#section-6.2.3.2).  For TLS 1.3, we would have to define a "randomized AEAD" whose output starts with N bytes of fresh entropy.

> I'm not sure whether we should require some mechanism for adding entropy to the inputs of encryption, particularly at this early stage. It might make sense to just leave this option in our back pocket for now.

I'm open to adding entropy instead of this PR, especially if it reduces the amount of specification text.  I'll give that a try in a separate PR for comparison.

Reply:
> I'm not aware of any well-known STPRPs that could be modeled this way, and I'm hesitant to start trying to classify them or invent one. Inventing a new AEAD (as here) actually seems easier (and more computationally efficient).

Oh, perhaps I've misunderstood what this PR is meant to do. My understanding is that all that is changing is we're adding an AEAD algorithm that would be used for the cTLS record layer. Does it also change how we handle plaintext records and record headers?

Reply:
This PR (#10) is about a new TLS 1.3 ciphersuite that would be negotiated and used instead of TLS_AES_128_GCM_SHA256.  It doesn't directly change anything about the Pseudorandom layer.

I've opened a new PR (#11) that tries to solve this differently, using extra entropy at the Pseudorandom layer.

Reply:
To be clear, I think the question of whether to add random padding to increase entropy is a separate question. Fundamentally what this PR is about is making the most of the entropy that we have. I've added a comment to #11 with details.